### PR TITLE
feat(replicated-library): add preflight pre-install hook

### PR DESCRIPTION
**Pull request by DexterYan**
This is another approach to add feature to [sc-75516](https://app.shortcut.com/replicated/story/75516/add-preflight-pre-install-hook-to-helm-library-chart)

### CODE REVIEW

**Comment by DexterYan** [comment]
## example
```
  preflights:
    my-preflights:
      image: replicated/preflight
      enabled: true
      collectors:
        - clusterInfo: {}
        - clusterResources: {}
      analyzers:
        - cephStatus: {}

```
## output
```
---
# Source: wg-easy/templates/replicated-library.yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: "test-preflight"
  annotations:
    "helm.sh/hook": pre-install, pre-upgrade
    "helm.sh/hook-weight": "-6"
    "helm.sh/hook-delete-policy": hook-succeeded, hook-failed
secrets:
- name: test-wg-easy-preflight-my-preflights
---
# Source: wg-easy/templates/replicated-library.yaml
apiVersion: v1
kind: Secret
metadata:
  name: test-wg-easy-preflight-my-preflights
  labels:
    troubleshoot.io/kind: preflight
    app.kubernetes.io/name: wg-easy
    app.kubernetes.io/instance=: test
    helm.sh/chart: wg-easy-0.1.0
  annotations:
    "helm.sh/hook": pre-install, pre-upgrade
    "helm.sh/hook-weight": "-6"
    "helm.sh/hook-delete-policy": before-hook-creation, hook-succeeded, hook-failed
stringData:
  preflight.yaml: |-
    apiVersion: troubleshoot.sh/v1beta2
    kind: Preflight
    metadata:
      name: test-wg-easy-preflight-my-preflights
    spec:
      collectors:      
        - clusterInfo: {}
        - clusterResources: {}
      analyzers:      
        - cephStatus: {}
---
# Source: wg-easy/templates/replicated-library.yaml
apiVersion: v1
kind: Pod
metadata:
 name: test-preflight-check
 labels:
    app.kubernetes.io/name: wg-easy
    app.kubernetes.io/instance=: test
    helm.sh/chart: wg-easy-0.1.0
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/version: "7"
 annotations:
   "helm.sh/hook": pre-install, pre-upgrade
   "helm.sh/show-output": "true"
   "helm.sh/hook-weight": "-5"
   "helm.sh/hook-delete-policy": before-hook-creation, hook-succeeded, hook-failed
   "helm.sh/hook-output-log-policy": hook-failed, hook-succeeded
spec:
 serviceAccountName: "test-preflight"
 restartPolicy: Never
 volumes:
  - name: preflights
    secret:
      secretName: test-wg-easy-preflight-my-preflights
  - name: kube-api-token
    projected:
      defaultMode: 420
      sources:
        - serviceAccountToken:
          expirationSeconds: 3607
          path: token
 containers:
  - name: pre-install-job
    image: "replicated/preflight"
    command:
      - "preflight"
      - "--interactive=false"
      - "/preflights/preflight.yaml"
    volumeMounts:
    - name: preflights
      mountPath: /preflights
```

**Comment by chris-sanders** [comment]
This looks really cool, FYI we're looking at moving some file paths https://github.com/replicatedhq/helm-charts/pull/80 this will impact where some of your files live. Just keep an eye on that PR.

File: charts/replicated-library/templates/lib/_preflights.tpl [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+      - "preflight"
+      - "--interactive=false"
+      - "/preflights/preflight.yaml"
+    volumeMounts:
+    - name: preflights
+      mountPath: /preflights
+
+---
+apiVersion: v1
+kind: ServiceAccount
```
Comment (xavpaice) [inline charts/replicated-library/templates/lib/_preflights.tpl:85]: suspect the service account, and RBAC roles, need to be setup using the library's RBAC tooling

**Comment by DexterYan** [comment]
if using 
```
preflights:
    my-preflights:
      image: replicated/preflight:latest
      enabled: true
      analyzers:
        - clusterVersion:
            outcomes:
              - fail:
                  when: "< 1.20.0"
                  message: This application requires at least Kubernetes 1.20.0, and recommends 1.22.0.
                  uri: https://www.kubernetes.io
              - warn:
                  when: "< 1.22.0"
                  message: Your cluster meets the minimum version of Kubernetes, but we recommend you update to 1.22.0 or later.
                  uri: https://kubernetes.io
              - pass:
                  when: ">= 1.22.0"
                  message: Your cluster meets the recommended and required versions of Kubernetes.
```
it is working as expected when the pod is running success.

However, if I am using
```
  preflights:
    my-preflights:
      image: replicated/preflight
      enabled: true
      collectors:
        - clusterInfo: {}
        - clusterResources: {}
```
The pod will fail, then the pod had been deleted automatically. However, sa, clusterrole, clusterrolebinding had not been deleted. Seems `hook-failed` has not trigger those delete

**Review by xavpaice — APPROVED** [review]

**Pull request merged**